Ensure header is visible even with large white images.

mikelehen opened the pull request:
Fixes https://app.asana.com/0/1204781676683007/1206803553364508 (once we pull this into the Shortwave build).

If an image would be too tall and overlap the header then we zoom out enough so that it doesn't.  The user can still zoom in via pinch-to-zoom, but that will hide the header (existing behavior) so that overlap isn't an issue.

# Before:
![image](https://github.com/shortwave/Lightbox/assets/206364/fbb7a7c1-a78a-4289-b5e8-56c689334f90)

# After:
<img width="394" alt="image" src="https://github.com/shortwave/Lightbox/assets/206364/053f334c-dbca-4d35-bab4-01b55550c785">

(and you can still hide the header by tapping or zooming)
![image](https://github.com/shortwave/Lightbox/assets/206364/7e0bf942-38a1-4610-b15f-534da13ea1f7)

mikelehen committed:
Ensure header is visible even with large white images.

Fixes https://app.asana.com/0/1204781676683007/1206803553364508 (once we pull this into the Shortwave build).

mikelehen commented on Source/LightboxController.swift:
```diff
@@ -210,7 +210,7 @@ open class LightboxController: UIViewController {
       y: 16,
       width: view.bounds.width,
-      height: 100
+      height: 45
```
100 was way bigger than necessary but it didn't matter since the header just overlayed the content.  Now we use this header size to adjust the actual content size (and avoid overlapping), so using a more appropriate value.

mikelehen commented:
@tanishqkancharla Feel like reviewing this one? Else I can send it to Jonny.

tanishqkancharla commented:
I'm happy to!

tanishqkancharla commented on Source/LightboxController.swift:
```diff
@@ -325,6 +325,16 @@ open class LightboxController: UIViewController {
 
+  fileprivate func calcScrollViewFrame(_ size: CGSize) -> CGRect {
+    let headerVisible = headerView.alpha == 1
+    let headerHeight = headerView.frame.minY + headerView.frame.height
```
I think this variable should be called headerMaxY and I believe you can just use headerView.frame.maxY

mikelehen replied:
Good suggestions, but n/a with new approach.

tanishqkancharla commented on Source/LightboxController.swift:
```diff
@@ -303,10 +303,10 @@ open class LightboxController: UIViewController {
     scrollView.contentSize = CGSize(
       width: size.width * CGFloat(numberOfPages) + spacing * CGFloat(numberOfPages - 1),
-      height: size.height)
+      height: scrollView.frame.height)
```
This doesn't seem right...why are we setting the contentSize equal to the height of the frame? Doesn't that mean it can't scroll or pan?

mikelehen replied:
I think it doesn't need to scroll/pan vertically, so it was okay?  But I may be confused about the different coordinate systems at play, especially when pinch-to-zoom comes into play. This might have been broken. 🤷

tanishqkancharla reviewed (commented):
Not sure why there isn't a code approve review attached but...

- What happens when you pinch to zoom? Does the header remain visible? In other apps (Discord), the header is only visible at the min zoom level
- Is the tap behavior default? I don't see any code related to it in this PR

mikelehen committed:
Make the header have a transparent background instead of resizing.

The previous approach broke pinch-to-zoom and was seeming complicated to fix. This is a different,
simpler, approach that justs adds a transparent background to the header so that the buttons are
visible even on a light background. It's not as nice, but works?

mikelehen reviewed (commented):
I don't think we have codeapprove hooked up to this repo. 🤷 

So it turns out I had broken pinch-to-zoom. 🤦   Thanks for flagging.  I messed around for a couple hours but couldn't get it to work smoothly.  Basically when we toggle the controls, we have to re-layout a bunch of stuff.  But when we re-layout, we reset the pinch-to-zoom.  And so it just ends up kinda' broken.

I gave up and did a lamer but easier approach of just giving the header a transparent black background so that even on a light background, the buttons will be visible.  I don't love it, but it solves the bug and I think is okay.  Updated the screenshot in the PR description to show it in action.

tanishqkancharla reviewed (commented):
Do you think we can get the discord behavior easily? Basically have top padding in the default state and the header is absolutely positioned. And zooming in or pinching in at all fades away the buttons.

tanishqkancharla reviewed (approved):
Otherwise, this behavior is totally fine. Up to you.

tanishqkancharla commented on Source/Views/HeaderView.swift:
```diff
@@ -67,9 +67,13 @@ open class HeaderView: UIView {
-    [closeButton, deleteButton].forEach { addSubview($0) }
+    let buttonView = UIView()
+    addSubview(buttonView)
+    buttonView.frame = CGRect(x: 0, y: 15, width: self.bounds.width, height: self.bounds.height)
```
I don't think I see why we need the extra "buttonView" subview here

mikelehen committed:
Zoom large images out initially to make room for header.

mikelehen commented on Source/LightboxController.swift:
```diff
@@ -208,9 +208,9 @@ open class LightboxController: UIViewController {
     headerView.frame = CGRect(
       x: 0,
-      y: 16,
+      y: 15,
```
I bumped this down to 15 basically just to make for more round numbers. Now the header uses up 60px and we make sure the images leave 120px padding by default (so when centered there's 60px for the header).

mikelehen commented on Source/Views/PageView.swift:
```diff
@@ -53,7 +53,7 @@ class PageView: UIScrollView {
 
   var hasZoomed: Bool {
-    return zoomScale != 1.0
+    return zoomScale != minimumZoomScale
```
This was a bug that now comes into play since we're tweaking the minimumZoomScale.

mikelehen commented on Source/Views/PageView.swift:
```diff
@@ -138,9 +138,8 @@ class PageView: UIScrollView {
-    let width = contentFrame.size.width / newZoomScale
-    let height = contentFrame.size.height / newZoomScale
+    let width = imageView.frame.width / newZoomScale
+    let height = imageView.frame.height / newZoomScale
```
We're zooming the imageView (which is smaller than the contentFrame unless the image happens to be the exact same size). This caused weirdness, especially with <1 zoom values that we're now using.

mikelehen committed:
tweak

mikelehen commented on Source/Views/PageView.swift:
```diff
@@ -170,19 +169,28 @@ class PageView: UIScrollView {
 
+    // Size the imageView so that it has the same aspect ratio as the image, while using the
+    // maximum possible width or height.
+    let newImageViewSize: CGSize
```
Just renamed this for clarity.

mikelehen reviewed (commented):
Thanks for the nudge. I think this is basically the Discord behavior now...  In particular, if the image is too tall and would overlap the header then we zoom it out so that it doesn't.  And then when you zoom in the existing behavior to hide the header kicks in (like Discord).

tanishqkancharla reviewed (changes requested):
I'm still noticing some issues:

- Zooming out beyond the edges of the image does not snap back to zoom scale 1 (it should)
- The header disappears completely for me and I can't get it back (maybe related to previous)
- nit but should probably look into: I'm seeing a warning in Xcode saying: "UIScrollView is ignoring an attempt to set zoomScale to a non-finite value: -inf"

mikelehen committed:
Fix issue with -inf minimumZoomScale.

When you close / reopen the lightbox for some reason imageView.frame ends up as (0,0) for one
render before it gets initialized properly. This was messing up our minimumZoomScale calculation to
be -inf, letting you size arbitrarily small and not resetting properly on double-click, etc.

mikelehen reviewed (commented):
@tanishqkancharla Thanks for catching that!  It turns out when you close / reopen the lightbox, the imageView.frame was sized as (0,0) initially which was messing up the logic and I was setting a `minimumZoomScale` of `-inf`.  I missed that because I didn't test closing / re-opening. 🤷‍♂️   

Anyway, when that happens it immediately re-renders with the correct size, so I just added a `maxHeight > 0` check which covers that case and is generally a useful defensive check.

So I _think_ it is working well now, though I don't 100% trust myself so if you're willing to test / sanity-check again, I'd really appreciate it.  Thanks for your diligence! 🙏

tanishqkancharla commented on Source/Views/PageView.swift:
```diff
@@ -170,19 +169,28 @@ class PageView: UIScrollView {
+    let maxHeight = imageViewSize.height - 120
+    if (maxHeight > 0 && imageView.frame.height > maxHeight) {
+      minimumZoomScale = maxHeight / imageView.frame.height
+      zoomScale = minimumZoomScale
```
Why are we "zooming out" instead of changing the bounds to account for the height? Semantically, I think the header/footer are essentially just some top/bottom padding in the default state.

tanishqkancharla reviewed (changes requested):
I'm still seeing an issue: every time you swipe to the long image (the one that says 300 x 600), and then swipe back, the default zoom level of the images gets smaller

![Screenshot 2024-07-09 at 11 48 15 AM](https://github.com/shortwave/Lightbox/assets/31808124/1241073f-d962-44bf-8ece-70a0f2d208ee)

mikelehen commented:
Closing in favor of #3.  Thanks @tanishqkancharla for your time on this. Sorry it was a deadend. 😢
